### ISSUE TITLE
Zignatures search during aaa gives wrong result

### ISSUE THREAD

**Issue by p371jah** [rizinorg/rizin#272]
Zignature search was tested on a binary builded from this source:
```c
#include <stdio.h>

int main() {
	char buf[5];
	FILE *f = fopen ("test", "r");
	fseek (f, 5, 1);
	fread (buf, 5, 1, f);
	fclose (f);
}
```
This is how `main` looks in the related object file. Ideally I want to see almost the same result after providing `rizin` appropriate zignatures of functions `fopen`, `fseek`, `fread` and `fclose` from `stdio` lib. 
```asm
[0x00000000]> pdf @ sym.main
            ;-- section..text_mn_4:
            ;-- .text$mn:
            ;-- $LN4:
/ 111: int sym.main (int argc, char **argv, char **envp);
|           ; var int64_t var_20h @ rsp+0x20
|           ; var int64_t var_28h @ rsp+0x28
|           ; arg int argc @ rdi
|           ; arg char **argv @ rsi
|           0x00000130      push rbx                                   ; [04] -r-x section size 111 named .text_mn_4
|           0x00000132      sub rsp, 0x30
|           0x00000136      mov rax, qword [sym.imp.__security_cookie] ; [0xfc0:8]=0
|           0x0000013d      xor rax, rsp
|           0x00000140      mov qword [var_28h], rax
|           0x00000145      lea rdx, sym.___C__01KDCPPGHE_r            ; 0xdf0 ; "r"; RELOC 32 ??_C@_01KDCPPGHE@r@ @ 0x00000df0
|           0x0000014c      lea rcx, sym.___C__04CEJDCDCH_test         ; 0xe00 ; "test"; RELOC 32 ??_C@_04CEJDCDCH@test@ @ 0x00000e00
|           0x00000153      call fopen
|           0x00000158      mov edx, 5
|           0x0000015d      mov rcx, rax
|           0x00000160      mov rbx, rax
|           0x00000163      lea r8d, [rdx - 4]
|           0x00000167      call fseek
|           0x0000016c      mov edx, 5
|           0x00000171      lea rcx, [var_20h]
|           0x00000176      mov r9, rbx
|           0x00000179      lea r8d, [rdx - 4]
|           0x0000017d      call fread
|           0x00000182      mov rcx, rbx
|           0x00000185      call fclose
|           0x0000018a      xor eax, eax
|           0x0000018c      mov rcx, qword [var_28h]
|           0x00000191      xor rcx, rsp
|           0x00000194      call __security_check_cookie
|           0x00000199      add rsp, 0x30
|           0x0000019d      pop rbx
\           0x0000019e      ret
```

`fopen` and `fseek` are just small wrappers around internal functions
```asm
/ 11: file*sym.fopen (const char *filename, const char *mode);
|       :   0x000236b0      mov r8d, 0x40                              ; sym..debug_S ; [31] -r-x section size 11 named .text_mn_31
\       `=< 0x000236b6      jmp sym.___common_fsopen_D__YAPEAU_iobuf__QEBD0H_Z; RELOC 32 ??$common_fsopen@D@@YAPEAU_iobuf@@QEBD0H@Z @ 0x00021d70
```

```asm
/ 8: int sym.fseek (FILE *stream, long offset, int whence);
| bp: 0 (vars 0, args 0)
| sp: 0 (vars 0, args 0)
| rg: 1 (vars 0, args 1)
|       :   0x00024680      movsxd rdx, edx                            ; whence ; [46] -r-x section size 8 named .text_mn_46
\       `=< 0x00024683      jmp sym._common_fseek__YAHV__crt_stdio_stream___JH_Z; RELOC 32 ?common_fseek@@YAHV__crt_stdio_stream@@_JH@Z @ 0x00022600
```

`fread` and `fclose` are big enough
```asm
/ 98: int sym.fclose (FILE *stream);
|           ; arg int64_t arg_8h @ rsp+0x30
|           ; arg int64_t arg_10h @ rsp+0x38
|           ; arg FILE *stream @ rdi
|           ; arg int64_t arg4 @ rcx
|           0x00022350      mov qword [arg_10h], rbx                   ; [23] -r-x section size 98 named .text_mn_23
|           0x00022355      mov qword [arg_8h], rcx                    ; arg4
|           0x0002235a      push rdi                                   ; stream
|           0x0002235b      sub rsp, 0x20
|           0x0002235f      mov rbx, rcx                               ; arg4
|           0x00022362      test rcx, rcx                              ; arg4
|       ,=< 0x00022365      jne 0x22385
|       |   0x00022367      call _errno
|       |   0x0002236c      mov dword [rax], 0x16                      ; [0x16:4]=-1
|       |   0x00022372      call _invalid_parameter_noinfo
|       |   ; CODE XREF from sym.fclose @ 0x22394
|      .--> 0x00022377      or eax, 0xffffffff                         ; -1
|      :|   ; CODE XREF from sym.fclose @ 0x223b0
|     .---> 0x0002237a      mov rbx, qword [arg_10h]
|     ::|   0x0002237f      add rsp, 0x20
|     ::|   0x00022383      pop rdi
|     ::|   0x00022384      ret
|     ::`-> 0x00022385      mov eax, dword [rcx + 0x14]                ; arg4
|     ::    0x00022388      shr eax, 0xc
|     ::    0x0002238b      test al, 1
|     ::,=< 0x0002238d      je 0x22396
|     ::|   0x0002238f      call ?__acrt_stdio_free_stream@@YAXV__crt_stdio_stream@@@Z
|     :`==< 0x00022394      jmp 0x22377
|     : `-> 0x00022396      call _lock_file
|     :     0x0002239b      nop
|     :     0x0002239c      mov rcx, rbx
|     :     0x0002239f      call sym._fclose_nolock                    ; RELOC 32 _fclose_nolock @ 0x00022070
|     :     0x000223a4      mov edi, eax
|     :     ;-- $LN10:
|     :     0x000223a6      mov rcx, rbx
|     :     0x000223a9      call _unlock_file
|     :     0x000223ae      mov eax, edi
\     `===< 0x000223b0      jmp 0x2237a
```

```asm
/ 29: size_t sym.fread (void *ptr, size_t size, size_t nmemb, FILE *stream);
|           ; var int64_t var_20h @ rsp+0x20
|           ; arg size_t nmemb @ rdx
|           ; arg int64_t arg5 @ r8
|           ; arg int64_t arg6 @ r9
|           0x00023200      sub rsp, 0x38                              ; [29] -r-x section size 29 named .text_mn_29
|           0x00023204      mov qword [var_20h], r9                    ; arg6
|           0x00023209      mov r9, r8                                 ; arg5
|           0x0002320c      mov r8, rdx                                ; nmemb
|           0x0002320f      or rdx, sym._comp.id_0x01056665            ; nmemb
|           0x00023213      call sym.fread_s                           ; RELOC 32 fread_s @ 0x000233c0
|           0x00023218      add rsp, 0x38
\           0x0002321c      ret
```

Here you can see `main` code after `zo` + `aaa`:
```asm
[0x140001394]> pdf @ main
            ; CALL XREF from sym.__acrt_uninitialize_tmpfile @ 0x14000131f
/ 111: int main (int argc, char **argv, char **envp);
|           ; var int64_t var_20h @ rsp+0x20
|           ; var int64_t var_28h @ rsp+0x28
|           0x140001010      push rbx
|           0x140001012      sub rsp, 0x30
|           0x140001016      mov rax, qword [0x140019008]              ; [0x140019008:8]=0x2b992ddfa232 ; "2\xa2\xdf-\x99+"
|           0x14000101d      xor rax, rsp
|           0x140001020      mov qword [var_28h], rax
|           0x140001025      lea rdx, [0x1400102d0]                    ; "r" ; uint32_t arg2
|           0x14000102c      lea rcx, str.test                         ; 0x1400102d4 ; "test" ; uint32_t arg1
|           0x140001033      call sym.___common_fsopen_D__YAPEAU_iobuf__QEBD0H_Z
|           0x140001038      mov edx, 5                                ; int64_t arg2
|           0x14000103d      mov rcx, rax                              ; int64_t arg1
|           0x140001040      mov rbx, rax
|           0x140001043      lea r8d, [rdx - 4]
|           0x140001047      call sym._common_fseek__YAHV__crt_stdio_stream___JH_Z
|           0x14000104c      mov edx, 5                                ; int64_t arg2
|           0x140001051      lea rcx, [var_20h]                        ; int64_t arg1
|           0x140001056      mov r9, rbx                               ; int64_t arg4
|           0x140001059      lea r8d, [rdx - 4]
|           0x14000105d      call sym._fread_nolock
|           0x140001062      mov rcx, rbx                              ; int64_t arg1
|           0x140001065      call sym._validate_and_update_state_at_beginning_of_format_character___positional_parameter_base_DV__string_output_adapter_D___crt_std
|           0x14000106a      xor eax, eax
|           0x14000106c      mov rcx, qword [var_28h]
|           0x140001071      xor rcx, rsp                              ; uint32_t arg1
|           0x140001074      call fcn.140001110
|           0x140001079      add rsp, 0x30
|           0x14000107d      pop rbx
\           0x14000107e      ret
```
What is wrong:
1. `fopen` and `fseek` got names of internal functions.  IMO it is not ok. Well I understand that these function are too small so simple byte search by mask gives many false positives but search algo should consider xrefs. And also it should detect internal function themselfs that is impossible for now (see https://github.com/radareorg/radare2/issues/16979)
2. `fread` got name of another function `_fread_nolock` which has the same zign bytes (see below). So xrefs issue again.
3. Not sure about `fclose` but seems that is related to xref too
4. Original mangled names are missing since rizin uses filtered unmagled names during zigns generating.

```asm
/ 29: sym._fread_nolock (int64_t arg3, int64_t arg5, int64_t arg6);
|           ; var int64_t var_20h @ rsp+0x20
|           ; arg int64_t arg3 @ rdx
|           ; arg int64_t arg5 @ r8
|           ; arg int64_t arg6 @ r9
|           0x00022400      sub rsp, 0x38                              ; [25] -r-x section size 29 named .text_mn_25
|           0x00022404      mov qword [var_20h], r9                    ; arg6
|           0x00022409      mov r9, r8                                 ; arg5
|           0x0002240c      mov r8, rdx                                ; arg3
|           0x0002240f      or rdx, sym._comp.id_0x01056665            ; arg3
|           0x00022413      call sym._fread_nolock_s                   ; RELOC 32 _fread_nolock_s @ 0x000225c0
|           0x00022418      add rsp, 0x38
\           0x0002241c      ret
```

Also check `pdf` output of `main` after `zo` + `z/` + `aa;aac;aar`:
```asm
[0x140001394]> pdf @ main
            ; CALL XREF from entry0 @ 0x14000131f
/ 111: int main (int argc, char **argv, char **envp);
|           ; var int64_t var_20h @ rsp+0x20
|           ; var int64_t var_28h @ rsp+0x28
|           0x140001010      push rbx
|           0x140001012      sub rsp, 0x30
|           0x140001016      mov rax, qword [0x140019008]              ; [0x140019008:8]=0x2b992ddfa232 ; "2\xa2\xdf-\x99+"
|           0x14000101d      xor rax, rsp
|           0x140001020      mov qword [var_28h], rax
|           0x140001025      lea rdx, [0x1400102d0]                    ; "r"
|           0x14000102c      lea rcx, str.test                         ; 0x1400102d4 ; "test"
|           0x140001033      call fcn.1400033cc
|           0x140001038      mov edx, 5
|           0x14000103d      mov rcx, rax
|           0x140001040      mov rbx, rax
|           0x140001043      lea r8d, [rdx - 4]
|           0x140001047      call fcn.140003938
|           0x14000104c      mov edx, 5
|           0x140001051      lea rcx, [var_20h]
|           0x140001056      mov r9, rbx
|           0x140001059      lea r8d, [rdx - 4]
|           0x14000105d      call sign.bytes.sym.fread_0
|           0x140001062      mov rcx, rbx
|           0x140001065      call sign.bytes.sym.fclose_0
|           0x14000106a      xor eax, eax
|           0x14000106c      mov rcx, qword [var_28h]
|           0x140001071      xor rcx, rsp
|           0x140001074      call fcn.140001110
|           0x140001079      add rsp, 0x30
|           0x14000107d      pop rbx
\           0x14000107e      ret
```
Ok, `fopen` and `fseek` are not detected but `fread` and `fclose` are ok (except for this `_0` suffix in their names).

**Drag and drop zip archives containing the Additional info here, don't use external services or link.**
Binary + object file + stdio zigns: [zign_workdir.zip](https://github.com/rizinorg/rizin/files/5744453/zign_workdir.zip)

Signatures are implemented in:
- `librz/core/cmd_zign.c`
- `librz/analysis/sign.c`

**Comment by p371jah** [comment]
Given issues prevent users from using zignature pack for MSVC runtime libs during automatic analysis. So in summary:
1. Even in case `stdio` lib there are too many false positives. I can't imagine what will happen in case of whole `ucrt` lib. 
2. Some functions aren't detected.
3. Original names are lost.

**Comment by XVilka** [comment]
Since recently @swoops worked on signatures improvements maybe can be interested in this.

**Comment by p371jah** [comment]
I have updated my previous comment. Well, in fact, we can create zignature pack, but in case of using it during auto-analysis it will confuse users rather then help them.

Also we can create zignature pack only for a given version of `ucrt`. So another issue is how to handle multiple implementations of function.

**Comment by XVilka** [comment]
I think detecting particular version of the function might even be useful. We should preserve this information in the comment to the function. Probably we want to introduce the concept of the "metasignature" that includes signatures of multiple function/library versions.

**Comment by swoops** [comment]
The signature matching alg does not keep track of what has been matched. If a match occurs, it keeps trying to match. This means a flag will be set for each signature that does match and the function will be renamed each time. The function ends up named as whatever the last match was.

Currently xrefs are not used in the code at all. Making use of them during the matching process would result in a lot of chicken and egg problems. Using it after the search is complete is a neat idea. You could keep track of all the functions that have multiple matches then use xrefs to settle what the function name should be. I am slowly working on a different r2 feature right now, but if I end up doing something like this, I will let you know.

**Comment by fangxlmr** [comment]
@Robot-RRR This issue is asigned to you as RSoC micro-task, so you could work on this from now on. Feel free to ask questions and send Pull Requests.

In case for anyone who is not aware about [RSoC](https://keenlab.tencent.com/zh/2021/03/04/RSoC-keenlab-2021/).

CC @XVilka

**Comment by Autosaida** [comment]
@fangxlmr Sorry, I just saw that I was assigned to this issue, but actually I plan to solve another issue(#417) under the Signatures label, is it OK? 
By the way, when I try to analyze the EXE file given by @pelijah under this issue, the `aaa` command will directly cause the crash. What's the problem? The git commit ID that I test is 175849813e043ab47c26059fdde093119cd222c6.

**Comment by XVilka** [comment]
Can confirm the crash:
```
[i] ℤ gdb rizin                                                                                                                                                                                                                   16:28:58 
Reading symbols from rizin...
(gdb) run main.exe
Starting program: /home/user/.local/bin/rizin main.exe
Missing separate debuginfos, use: dnf debuginfo-install glibc-2.32-4.fc33.x86_64
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
 -- Use 'e asm.offset=true' to show offsets in 16bit segment addressing mode.
[0x140001394]> aaa
[x] Analyze all flags starting with sym. and entry0 (aa)
[x] Analyze function calls (aac)
[x] Analyze len bytes of instructions for references (aar)
[WARNING: rz_vector_empty: assertion 'vec' failed (line 74)

Program received signal SIGSEGV, Segmentation fault.
rz_analysis_class_method_exists_by_addr (analysis=<optimized out>, class_name=class_name@entry=0xe8e2e0 "", addr=5368719132) at ../librz/include/rz_vector.h:74
74              rz_return_val_if_fail(vec, false);
Missing separate debuginfos, use: dnf debuginfo-install bzip2-libs-1.0.8-4.fc33.x86_64 libgcc-10.2.1-9.fc33.x86_64 libstdc++-10.2.1-9.fc33.x86_64 xz-libs-5.2.5-4.fc33.x86_64
(gdb) bt
#0  rz_analysis_class_method_exists_by_addr (analysis=<optimized out>, class_name=class_name@entry=0xe8e2e0 "", addr=5368719132) at ../librz/include/rz_vector.h:74
#1  0x00007ffff6d90c7a in recovery_apply_vtable (context=0x7fffffffceb0, class_name=class_name@entry=0xe8e2e0 "", vtable_info=0xe8e2a0) at ../librz/analysis/rtti_msvc.c:835
#2  0x00007ffff6d918aa in recovery_apply_vtable (vtable_info=<optimized out>, class_name=0xe8e2e0 "", context=<optimized out>) at ../librz/analysis/rtti_msvc.c:818
#3  recovery_apply_complete_object_locator (col=0xe8e080, context=0x7fffffffcdc0) at ../librz/analysis/rtti_msvc.c:937
#4  recovery_apply_complete_object_locator (context=0x7fffffffcdc0, col=0xe8e080) at ../librz/analysis/rtti_msvc.c:897
#5  0x00007ffff6d927d1 in rz_analysis_rtti_msvc_recover_all (vt_context=<optimized out>, vtables=0x7fffffffcdc0) at ../librz/analysis/rtti_msvc.c:1021
#6  0x00007ffff6d9073d in rz_analysis_rtti_recover_all (analysis=<optimized out>) at ../librz/analysis/rtti.c:95
#7  0x00007ffff6d72fd2 in rz_analysis_class_recover_all (analysis=<optimized out>) at ../librz/analysis/class.c:1445
#8  0x00007ffff694feec in rz_core_analysis_everything (core=0x7ffff679e010, experimental=<optimized out>, dh_orig=<optimized out>) at ../librz/core/canalysis.c:7078
#9  0x00007ffff69e1681 in cmd_analysis_all (input=0x700312 "a", core=0x7ffff679e010) at ../librz/core/cmd_analysis.c:8155
#10 rz_cmd_analysis (data=0x7ffff679e010, input=0x700311 "aa") at ../librz/core/cmd_analysis.c:9008
#11 0x00007ffff69f6a53 in call_cd (args=<optimized out>, cd=0x4640d0, cmd=<optimized out>) at ../librz/core/cmd_api.c:654
#12 rz_cmd_call_parsed_args (cmd=0x45bd90, args=args@entry=0x629390) at ../librz/core/cmd_api.c:669
#13 0x00007ffff69c9212 in handle_ts_arged_command_internal (node_string=0x55ba70 "aaa", node=..., state=0x7fffffffd2c0) at ../librz/core/cmd.c:4625
#14 handle_ts_arged_command (state=0x7fffffffd2c0, node=...) at ../librz/core/cmd.c:4587
#15 0x00007ffff698a230 in handle_ts_command (state=state@entry=0x7fffffffd2c0, node=...) at ../librz/core/cmd.c:6056
#16 0x00007ffff69ba2d7 in handle_ts_commands_internal (node_string=0x422c00 "aaa", node=..., state=0x7fffffffd2c0) at ../librz/core/cmd.c:6113
#17 handle_ts_commands (state=state@entry=0x7fffffffd2c0, node=...) at ../librz/core/cmd.c:6078
#18 0x00007ffff69ba6a7 in core_cmd_tsr2cmd (core=<optimized out>, cstr=0x408dc0 "aaa", split_lines=split_lines@entry=false, log=log@entry=true) at ../librz/core/cmd.c:6215
#19 0x00007ffff69adfbe in rz_core_cmd (core=core@entry=0x7ffff679e010, cstr=0x408dc0 "aaa", log=log@entry=1) at ../librz/core/cmd.c:6262
#20 0x00007ffff69fa70b in rz_core_prompt_exec (r=r@entry=0x7ffff679e010) at ../librz/core/core.c:2811
#21 0x00007ffff69fada8 in rz_core_prompt_loop (r=r@entry=0x7ffff679e010) at ../librz/core/core.c:2662
#22 0x00007ffff7ec1a78 in rz_main_rizin (argc=<optimized out>, argv=<optimized out>) at ../librz/main/rizin.c:1352
#23 0x00007ffff7ce71e2 in __libc_start_main () from /lib64/libc.so.6
#24 0x000000000040148e in _start ()
```

**Comment by fangxlmr** [comment]
> @fangxlmr Sorry, I just saw that I was assigned to this issue, but actually I plan to solve another issue(#417) under the Signatures label, is it OK?
> By the way, when I try to analyze the EXE file given by @pelijah under this issue, the `aaa` command will directly cause the crash. What's the problem? The git commit ID that I test is [1758498](https://github.com/rizinorg/rizin/commit/175849813e043ab47c26059fdde093119cd222c6).

@Robot-RRR Sure.

**Comment by fangxlmr** [comment]
Likewise, @chinggg This issue is asigned to you as RSoC micro-task, so you could work on this from now on. Feel free to ask questions and send Pull Requests.

In case for anyone who is not aware about [RSoC](https://keenlab.tencent.com/zh/2021/03/04/RSoC-keenlab-2021/).

CC @XVilka

(And bear with me here for adding similar comments related to RSoC, THANKS!)

**Comment by chinggg** [comment]
Compared with last micro-task, this one seems to be a tough nut to crack :crying_cat_face: 
Sadly, the `aaa` command still causes the crash with the lastest commit. It seems I should try to find the reason of the crash and fix it before researching on the improvement of signature matching? :thinking: 

@XVilka Do you have any suggestions? Thanks!

**Comment by p371jah** [comment]
@chinggg I have fixed that bug.

**Comment by chinggg** [comment]
@pelijah Hi, I want to know how `ucrt_stdio.sdb` is created. Do you create it by manually signing the `main.obj` file? Since the function names are long and unofficial. If so, we might not expect the signature to work for all MSVC executable files using `stdio`. That might need machine learning.
And now the problem is signature search gives wrong result even the `ucrt_stdio.sdb` shares the same origin with `main.exe`? So the algo to search and match the signature must be improved, and that is all the work. Do I understand it correctly? Thanks!

@swoops Hi, do you have any progress on this feature of radare2? Thanks!

**Comment by p371jah** [comment]
`ucrt_stdio.sdb` was generated from object files of ucrt lib from Windows SDK. `main.exe` was linked against a bit older version of `ucrt` lib but I suppose that implementation of `stdio` functions haven't changed.

**Comment by swoops** [comment]
> @swoops Hi, do you have any progress on this feature of radare2? Thanks!

I have just been thinking on this a lot, no code so far. If I get some actual code working I will ping here. I understand if you don't want to wait on me.

**Comment by chinggg** [comment]
~~I found `zb` has no output at `0x1400033cc` but it should print `sym.___common_fsopen_D__YAPEAU_iobuf__QEBD0H_Z` just as `zb` prints `sym._fread_nolock_s` at `0x1400033d8`.
It is weird that `zb` cannot find match zignature at current offset even I found the bytes at that offset are nearly the same as `sym.fopen` in the zignature.~~

I used `gdb` and found `zb` at `0x1400033cc` will not call `rz_sign_find_closest_sig()`, that is because `do_bestmatch_sig()` return false immediately if function signature size is smaller than minsz. There should be warning about this but a `\n` is missing in `eprintf()` so the warn is not shown. I have fixed that in #1096 

And `zbr sym.fopen` has the output `1.00000 B   sym.___common_fsopen_D__YAPEAU_iobuf__QEBD0H_Z`, which means rizin has found the best matched function for the signature `sym.fopen` even it has wrong name. But in radare2, `zbr` cannot print correct anwser for `sym.fopen` and `sym.fseek`, which is worse than Rizin.

So it seems searching the whole file for specific signature works fine but finding correspoinding zignature at specific offset fails for `fopen` and `fseek`, which are both short wrapper function. I think if I can fix `zb` to print the best matched zigname, then the wrong result in during analysis can be solved similarly. 🤔

**Comment by chinggg** [comment]
@swoops Hi, I found your `zb` command do the similar thing (find closest matching zignatures to function at current offset) as zignature search during `aaa`, but they use different functions thus giving different results. `aaa` will call `search()` in `cmd_zign.c` but `zb` called `do_bestmatch_sig()` which in turn calls `rz_sign_find_closest_sig()` to calculate scores and return list of possible results.

It seems result of `zb` is more correct than `aaa`. And I do not understand the difference between the two approaches now, is it possible to let `aaa` call `do_best_match_sig()` instead of `search()` which was written 3 years ago?

And `z.` has no hits at `0x1400033cc`, that may be related to some threshold things.

In summary, I will have to improve function `search()` or try to reuse the `zb` results to serve `aaa` command. 😺

**Comment by swoops** [comment]
I intended `zb` to be for manual testing when `z/` failed. It will be slower and it will always find a result unless threshold is set. I think using the `do_bestmatch_sig` in `aaa` would result in more matches but also more false positives.

Thanks for finding that missing `\n`.

**Comment by swoops** [comment]
https://github.com/radareorg/radare2/pull/18692 is my first step toward preventing miss-naming functions in r2.

**Comment by ret2libc** [comment]
As this is not a regression and we are focused on other things for the upcoming 0.3.0 release, I'm moving this for next release. We'll try to focus on the whole zignature thing a bit more then.

**Comment by XVilka** [comment]
I does work with FLIRT signatures, and old ones will be removed in 0.5.0 release. Thus moving the milestone, but not yet closing until removal is done.
